Add plugin: Varinote

gsarig opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin:

## Release Checklist
- [ ] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [ ]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

Zachatoo reviewed (commented):
Remade this comment below.

joethei commented:
Bad bot

Zachatoo commented:
Re-making my review comment here to try to fix the bot.

[this.app.workspace.on('file-open', this.onFileOpen);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L29), [this.app.workspace.off('layout-ready', this.onLayoutReady);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L34-L35)
Prefer to use `registerEvent()` instead of manually attaching and detaching the event listener. Events need to be registered, so they can be properly unloaded when the plugin is disabled. <https://docs.obsidian.md/Plugins/Events>

[this.app.vault.read(fileToCheck).then((content: string) => {](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/utils/processActiveFile.ts#L19)
[Prefer async/await over Promise](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer%20async/await%20over%20Promise)

[if (this.app.workspace.layoutReady) {](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L24-L26)
For users that don't have the core [Templates plugin](https://help.obsidian.md/Plugins/Templates) enabled, this is going to execute template code in the active note when Obsidian is restarted or your plugin is enabled/updated. Is that intentional? I would recommend not processing the active file in your `onload` function.

[this.app.workspace.on('file-open', this.onFileOpen);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L29)
For users that don't have the core [Templates plugin](https://help.obsidian.md/Plugins/Templates) enabled, this is going to execute template code when switching between notes. Is that intentional? I would recommend listening to the [Vault.on('create')](https://docs.obsidian.md/Reference/TypeScript+API/Vault/on('create')) event instead.

[const content = await this.app.vault.read(file);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/utils/triggerModal.ts#L12-L16)
[Prefer Vault.process instead of Vault.modify to modify a file in the background](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer%20%60Vault.process%60%20instead%20of%20%60Vault.modify%60%20to%20modify%20a%20file%20in%20the%20background)

[new Setting(containerEl).setName('Varinote (Variables in Notes)').setHeading();](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/settings/Settings.tsx#L17)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

[.addSlider(slider => {](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/fields/sliderField.ts#L42-L50)
It would be a nice touch to use [setDynamicTooltip](https://docs.obsidian.md/Reference/TypeScript+API/SliderComponent/setDynamicTooltip) to show the number when hovering over or changing the slider value.

[dropdown.setValue(firstOption);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/fields/dropdownField.ts#L37)
If a user doesn't touch given dropdown in the modal, all of the options for that dropdown will be inserted instead of just the selected option. I would expect only the the selected dropdown option to be inserted.

gsarig commented:
Hi @Zachatoo , and thank you for the great feedback!

> Re-making my review comment here to try to fix the bot.
> 
> [this.app.workspace.on('file-open', this.onFileOpen);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L29), [this.app.workspace.off('layout-ready', this.onLayoutReady);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L34-L35) Prefer to use `registerEvent()` instead of manually attaching and detaching the event listener. Events need to be registered, so they can be properly unloaded when the plugin is disabled. https://docs.obsidian.md/Plugins/Events
> 

Done!

> [this.app.vault.read(fileToCheck).then((content: string) => {](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/utils/processActiveFile.ts#L19) [Prefer async/await over Promise](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer%20async/await%20over%20Promise)
> 

Done!

> [if (this.app.workspace.layoutReady) {](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L24-L26) For users that don't have the core [Templates plugin](https://help.obsidian.md/Plugins/Templates) enabled, this is going to execute template code in the active note when Obsidian is restarted or your plugin is enabled/updated. Is that intentional? I would recommend not processing the active file in your `onload` function.
> 
> [this.app.workspace.on('file-open', this.onFileOpen);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/main.ts#L29) For users that don't have the core [Templates plugin](https://help.obsidian.md/Plugins/Templates) enabled, this is going to execute template code when switching between notes. Is that intentional? I would recommend listening to the [Vault.on('create')](https://docs.obsidian.md/Reference/TypeScript+API/Vault/on('create')) event instead.
> 

Can you give me an example of the issue? The plugin isn't dependent on Templates or Templater, and could, potentially, work standalone. For example, the user could paste a Varinote template like for example:

````markdown
```varinote 

var_1:: Label 1|Value 1
var_2:: Label 2

```

## My text with variables:

* This is the value of my first variable: {{$var_1}}
* This is the value of the second: {{$var_2}}.

````

anywhere in the note, and then trigger the `Varinote` command from the command palette. This would trigger the modal, and when the user entered the values, it would transform the content. This is the intended behavior unless I misunderstood the issue. 

> [const content = await this.app.vault.read(file);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/utils/triggerModal.ts#L12-L16) [Prefer Vault.process instead of Vault.modify to modify a file in the background](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer%20%60Vault.process%60%20instead%20of%20%60Vault.modify%60%20to%20modify%20a%20file%20in%20the%20background)
> 

Done!

> [new Setting(containerEl).setName('Varinote (Variables in Notes)').setHeading();](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/settings/Settings.tsx#L17) [Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)
> 

Done! 

> [.addSlider(slider => {](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/fields/sliderField.ts#L42-L50) It would be a nice touch to use [setDynamicTooltip](https://docs.obsidian.md/Reference/TypeScript+API/SliderComponent/setDynamicTooltip) to show the number when hovering over or changing the slider value.
> 

I know :) This was in my to-do list, because it wasn't so user-friendly. Unfortunately, `setDynamicTooltip` didn't work as expected, so I made a custom implementation. I hope that it is OK with you. 

> [dropdown.setValue(firstOption);](https://github.com/gsarig/obsidian-varinote/blob/6993021a8eb5fa9d0a3ff1c9884649d590fd7774/src/fields/dropdownField.ts#L37) If a user doesn't touch given dropdown in the modal, all of the options for that dropdown will be inserted instead of just the selected option. I would expect only the the selected dropdown option to be inserted.

That's a great catch, thanks! I've included a fix and should work now. 

If it helps, you can see all the changes of the latest release on this [PR](https://github.com/gsarig/obsidian-varinote/pull/2/files).

Please let me know if you need anything else.

Thanks!

Zachatoo commented:
@gsarig Here's a video example of what I mean by the template code being executed when restarting Obsidian, updating the plugin, or switching between notes.

https://github.com/user-attachments/assets/8b44215d-9f52-4919-8897-97c8107c11ed

[await this.app.vault.process(file, () => updatedContent);](https://github.com/gsarig/obsidian-varinote/blob/d0496eb2afff549f13022c1ae57c8fd7b34d0388/src/utils/triggerModal.ts#L33)
This is not the correct way to safely make atomic changes, this is the equivalent of what you were doing before, except with an additional unnecessary file read. This video from the Obsidian team has a deep dive on how to use the `Vault.process` API correctly. https://youtu.be/RvtV9DKr3ag?t=52

Thanks for making those other changes!

gsarig commented:
@Zachatoo Thanks for the clarifications and your patience on this :) 


> @gsarig Here's a video example of what I mean by the template code being executed when restarting Obsidian, updating the plugin, or switching between notes.
> 
>  2024-12-07_13-45-50.mp4 

I see now what you mean. I noticed this behavior too, but I didn't consider it a problem. I can see though now why you do, and agree that it can be confusing. While fixing the Modal display when switching notes was easy, I am not sure what's the best way to do it to prevent it when the vault opens. I made a change that is working for me and you can see it [here](https://github.com/gsarig/obsidian-varinote/pull/3/commits/493a6dc455fcea962b37cdb731c27b07ccb92695). As you can see, I capture the current session time, and use it to only process the files created after the plugin loading time. If this is an acceptable solution, let me know so that I push the new release. 

> [await this.app.vault.process(file, () => updatedContent);](https://github.com/gsarig/obsidian-varinote/blob/d0496eb2afff549f13022c1ae57c8fd7b34d0388/src/utils/triggerModal.ts#L33) This is not the correct way to safely make atomic changes, this is the equivalent of what you were doing before, except with an additional unnecessary file read. This video from the Obsidian team has a deep dive on how to use the `Vault.process` API correctly. https://youtu.be/RvtV9DKr3ag?t=52
> 

I think I get it now. I've implemented this [here](https://github.com/gsarig/obsidian-varinote/pull/3/commits/a8b64655ee0804c637f6cf9984164bae55949865), following the example from the video. Is this what you would expect?

Zachatoo commented:
> While fixing the Modal display when switching notes was easy, I am not sure what's the best way to do it to prevent it when the vault opens.

You can use [`Workspace.onLayoutReady()`](https://docs.obsidian.md/Reference/TypeScript+API/Workspace/onLayoutReady). For example:
```js
this.app.workspace.onLayoutReady(() => {
  this.registerEvent(this.app.vault.on('create', (file: TAbstractFile) => {
    this.handleFileCreate(file);
  }));
});
```

The process change you've made is correct, nice job!

gsarig commented:
> > While fixing the Modal display when switching notes was easy, I am not sure what's the best way to do it to prevent it when the vault opens.
> 
> You can use [`Workspace.onLayoutReady()`](https://docs.obsidian.md/Reference/TypeScript+API/Workspace/onLayoutReady). For example:
> 
> ```js
> this.app.workspace.onLayoutReady(() => {
>   this.registerEvent(this.app.vault.on('create', (file: TAbstractFile) => {
>     this.handleFileCreate(file);
>   }));
> });
> ```

Thanks @Zachatoo I've applied your suggested and now it is ready for a re-check.

joethei commented:
[Copyright (c) 2023 xRyul](https://github.com/gsarig/obsidian-varinote/blob/9af14ccf347e6306a2da2143c6bd24e925bf9165/LICENSE#L3)
Update this.

["description": "Add variables in Templates and set their values on-the-fly during the Note creation.",](https://github.com/gsarig/obsidian-varinote/blob/9af14ccf347e6306a2da2143c6bd24e925bf9165/manifest.json#L6)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[export function createTooltip(container: HTMLElement, initialValue: number): HTMLElement {](https://github.com/gsarig/obsidian-varinote/blob/9af14ccf347e6306a2da2143c6bd24e925bf9165/src/utils/getTooltip.ts#L1-L11)
The Obsidian API exports the `setTooltip` function for this purpose.

[this.addSettingTab(new VarinoteSettings(this.app, this));](https://github.com/gsarig/obsidian-varinote/blob/9af14ccf347e6306a2da2143c6bd24e925bf9165/src/main.ts#L14)
Your settings tab contains no actual settings, so it should not be added.

[containerEl.createEl('p', {](https://github.com/gsarig/obsidian-varinote/blob/9af14ccf347e6306a2da2143c6bd24e925bf9165/src/settings/Settings.tsx#L17-L19)
For when you have settings:
[Don't add a top-level heading / text in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.)

gsarig commented:
Hi @joethei Thanks for the feedback. I've addressed the issues that you flagged, except that one: 

>
> [export function createTooltip(container: HTMLElement, initialValue: number): HTMLElement {](https://github.com/gsarig/obsidian-varinote/blob/9af14ccf347e6306a2da2143c6bd24e925bf9165/src/utils/getTooltip.ts#L1-L11) The Obsidian API exports the `setTooltip` function for this purpose.
> 


For that one, I already tried to use `setDynamicTooltip` a few weeks ago, but the problem was that I wanted to have the value always visible on a fixed position:

![image](https://github.com/user-attachments/assets/c4e453b3-e0af-4b45-af3f-483bea2d3901)

When I tried it, I noticed 2 problems for what I wanted to do:

1. `setDynamicTooltip` would append the tooltip on mouseover but would remove it on mouseout, so I couldn't keep it on a fixed position with CSS.
2. If I remember correctly (it's been a while since I tried it), even on mouseover, it wouldn't follow the slider, which would make more sense for the way that it worked. 

The most important reason that I didn't go with it, though, was the first, as I wanted to show the value all the time. 

Is this OK with you or should  I try a different approach?

joethei commented:
[Copyright (c) 2023 gsarig](https://github.com/gsarig/obsidian-varinote/blob/4c851ceb1115a5d65d045d1598c9d80704e61aa4/LICENSE#L3)
Still using the wrong year.

gsarig commented:
> [Copyright (c) 2023 gsarig](https://github.com/gsarig/obsidian-varinote/blob/4c851ceb1115a5d65d045d1598c9d80704e61aa4/LICENSE#L3) Still using the wrong year.

@joethei Updated
